src/main/java/com/hpe/adm/octane/ideplugins/services/connection/IdePluginsOctaneHttpClient.java

```diff
@@ -0,0 +1,628 @@
+			switch (octaneHttpRequest.getOctaneRequestMethod()) {
+			case GET: {
+				GenericUrl domain = new GenericUrl(octaneHttpRequest.getRequestUrl());
+				httpRequest = requestFactory.buildGetRequest(domain);
+				httpRequest.getHeaders().setAccept(((OctaneHttpRequest.GetOctaneHttpRequest) octaneHttpRequest).getAcceptType());
+				final String eTagHeader = requestToEtagMap.get(octaneHttpRequest);
+				if (eTagHeader != null) {
+					httpRequest.getHeaders().setIfNoneMatch(eTagHeader);
+				}
+				break;
+			}
+			case POST: {
+				OctaneHttpRequest.PostOctaneHttpRequest postOctaneHttpRequest = (OctaneHttpRequest.PostOctaneHttpRequest) octaneHttpRequest;
+				GenericUrl domain = new GenericUrl(octaneHttpRequest.getRequestUrl());
+				httpRequest = requestFactory.buildPostRequest(domain, ByteArrayContent.fromString(null, postOctaneHttpRequest.getContent()));
+				httpRequest.getHeaders().setAccept(postOctaneHttpRequest.getAcceptType());
+				httpRequest.getHeaders().setContentType(postOctaneHttpRequest.getContentType());
+				break;
+			}
+			case POST_BINARY: {
+				httpRequest = buildBinaryPostRequest((OctaneHttpRequest.PostBinaryOctaneHttpRequest) octaneHttpRequest);
+				break;
+			}
+			case PUT: {
+				OctaneHttpRequest.PutOctaneHttpRequest putHttpOctaneHttpRequest = (OctaneHttpRequest.PutOctaneHttpRequest) octaneHttpRequest;
+				GenericUrl domain = new GenericUrl(octaneHttpRequest.getRequestUrl());
+				httpRequest = requestFactory.buildPutRequest(domain, ByteArrayContent.fromString(null, putHttpOctaneHttpRequest.getContent()));
+				httpRequest.getHeaders().setAccept(putHttpOctaneHttpRequest.getAcceptType());
+				httpRequest.getHeaders().setContentType(putHttpOctaneHttpRequest.getContentType());
+				break;
+			}
+			case DELETE: {
+				GenericUrl domain = new GenericUrl(octaneHttpRequest.getRequestUrl());
+				httpRequest = requestFactory.buildDeleteRequest(domain);
+				break;
+			}
+			default: {
+				throw new IllegalArgumentException("Request method not known!");
+			}
+			}
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+		return httpRequest;
+	}
+
+	/**
+	 * Convert google implementation of {@link HttpResponse} to an implementation abstract {@link OctaneHttpResponse}
+	 * @param httpResponse implementation specific {@link HttpResponse}
+	 * @return {@link OctaneHttpResponse} created from the impl response object
+	 */
+	protected OctaneHttpResponse convertHttpResponseToOctaneHttpResponse(HttpResponse httpResponse) {
+		try {
+			return new OctaneHttpResponse(httpResponse.getStatusCode(), httpResponse.getContent(), httpResponse.getContentCharset());
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	@Override
+	public OctaneHttpResponse execute(OctaneHttpRequest octaneHttpRequest) {
+		//Do not execture a http request while the authentication method is running
+		authenticationLock.lock();
+		authenticationLock.unlock();
+		return execute(octaneHttpRequest, HTTP_REQUEST_RETRY_COUNT);
```

Comment:
As i understand here you request a lock which blocks this thread, later on somewhere else this lock is released and the thread continues so the unlock is going to be redundant, i would like to know how this works.

Reply:
authenticationLock is locked during authentication (authenticate method)
there it locks at the beginning and unlocks at the end.
the issue is that no request should be allowed to execute while the authentication is going on, i.e. the user is logging into the embedded browser.
trying to get the same lock in the generic req execute means that all generic req wait until auth is done, but then they immediately release their lock, because we don't want to make the http client single threaded.
the only thing we want to prevent is executing any request while the authenticate method is running

Reply:
I think i understood now, you're consuming the lock so nothing else can consume it at that moment. Right?